staging/src/k8s.io/client-go/tools/cache/reflector_test.go

```diff
@@ -105,6 +106,54 @@ func TestRunUntil(t *testing.T) {
 	}
 }
 
+// TestReflectorRunWithStopOptions tests that when the lister is forced to always error out
+// the reflector will stop when it's StopOnListError returns true
+// and the reflector will not stop when it's StopOnListError returns false.
+func TestReflectorRunWithStopOptions(t *testing.T) {
+	table := []struct {
+		stopOnListError bool
+	}{
+		{true},
+		{false},
+	}
+	for _, item := range table {
+
+		store := NewStore(MetaNamespaceKeyFunc)
+		r := NewReflector(&testLW{}, &v1.Pod{}, store, 0)
+		listErr := errors.New("list error")
+		_ = watch.NewFake()
+		r.listerWatcher = &testLW{
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				panic("unreachable")
+			},
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return nil, listErr
+			},
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+		go func() {
+			// confirm the reflector stops running when it hits a list error
+			defer cancel()
+			r.RunWithStopOptions(ctx, StopOptions{
+				StopOnListError: func(err error) bool {
+					return item.stopOnListError
+				},
+			})
+		}()
+		select {
+		case <-ctx.Done():
+			if !item.stopOnListError {
+				t.Errorf("reflector should NOT have stopped when stopOnListError is false")
+			}
+		case <-time.After(3 * time.Second):
```

Comment:
I'm not sure we want to use that here, because we actually are going to wait this long for the test to pass. We are testing both when stopOnListError is true and false, so in the false condition, it will not stop and we expect it to timeout after the desired wait time. I'm open to better suggestions about what to use as the wait time, it just needs to be long enough to not fire before stopOnListError does. 

I tried injecting a fake clock into the reflector, but the issue is that you still need to block some amount of time in order for the go routine running the reflector to start it and recognize the list error. 

I actually think it can be much less than 3 seconds though, so in order to avoid slowing down the test too much, I've bumped it down to 200ms for both here and in the shared_informer test. It really only needs to wait for as long as it takes the reflector to make its inital list call and error out.

Let me know if you think there is a better approach.

Reply:
the timeout cannot be too short, but it can be arbitrarily long, right? i think the best way is to run the test 10000 times under stress, varying timeout to see what is the best timeout value. i think a few seconds are fine, the tests are run concurrently, the timeout will slow down this one test, but it won't affect other tests.

go test k8s.io/client-go/tools/cache --race --count 10000
or use https://pkg.go.dev/golang.org/x/tools/cmd/stress

Reply:
Cool, thanks for the advice. I ran it with stress and started seeing about 1 failure every couple thousand runs at 10ms. 20ms hit an error at 16k runs, and 100ms seemed to go fine fora as long as I let the stress test run for (100k+ tests).

I've added a comment explaining this.

If you'd feel more comfortable with a higher wait I can bump it up higher too.